students/yushus/session07/html_render.py

```diff
@@ -0,0 +1,123 @@
+HTML Renderer assignment
+"""
+
+import os
+import random
+import re
+import sys
+
+from collections import defaultdict
+
+# This is the framework for the base class
+class Element(object):
+    tag = "html"
+    indent = "    "
+
+    def __init__(self, content=None, **attributes):
+        self.contents = [content] if content else []
+        self.attributes = attributes
+    
+    def _open_tag(self):
+        return f"<{self.tag}{self._get_attributes()}>"
+
+    def _close_tag(self):
+        return f"</{self.tag}>"
+
+    def _get_attributes(self):
+        page_attributes = ""
+        if self.attributes:
+            for attribute in self.attributes:
+                page_attributes = f"{page_attributes} {attribute}=\"{self.attributes[attribute]}\""
+        return page_attributes
+
+    def render(self, out_file, cur_ind=""):
+        out_file.write(cur_ind + self._open_tag())
+        out_file.write("\n")
+
+        for content in self.contents:
+            if isinstance(content, Element):
+                content.render(out_file, cur_ind + self.indent)
+            elif isinstance(content, str):
+                out_file.write(f"{cur_ind}{self.indent}{content}")
+                out_file.write("\n")
+            else:
+                raise TypeError("Cannot render!!")
+
+        out_file.write(cur_ind + self._close_tag())
+        out_file.write("\n")
+
+    def append(self, new_content):
+        self.contents.append(new_content)
+
+class Html(Element):
+    tag = "html"
+    def render(self, out_file, cur_ind=""):
+        out_file.write(f"{cur_ind}<!DOCTYPE {self.tag}>")
+        out_file.write("\n")
+        super().render(out_file, cur_ind)
+
+class Body(Element):
+    tag = "body"
+
+class P(Element):
+    tag = "p"
+
+class Head(Element):
+    tag = "head"
+
+class OneLineTag(Element):
+    def append(self, content):
+        raise NotImplementedError
+
+    def render(self, out_file, cur_ind=""):
+        out_file.write(cur_ind + super()._open_tag())
+        out_file.write(self.contents[0])
+        out_file.write(super()._close_tag())
+        out_file.write("\n")
+
+class Title(OneLineTag):
+    tag = "title"
+
+class SelfClosingTag(Element):
+    def __init__(self, content=None, **attributes):
+        if content is not None:
+            raise TypeError("SelfClosingTag can not contain any content")
+        super().__init__(content=content, **attributes)
+
+    def render(self, outfile, cur_ind=""):
+        tag = super()._open_tag()[:-1] + " />\n"
+        outfile.write(f"{cur_ind}{tag}")
+
+    def append(self, *args):
+        raise TypeError("You can not add content to a SelfClosingTag")
+
+class Hr(SelfClosingTag):
+    tag = "hr"
+
+class Br(SelfClosingTag):
+    tag = "br"
+
+class Meta(SelfClosingTag):
+    tag  = "meta"
+
+class A(OneLineTag):
+    tag = "a"
+    def __init__(self, link, content=None, **attributes):
+        attributes["href"] = link
+        super().__init__(content, **attributes)
```

Comment:
Good job, reusing the superclass as much as possible! This way, you didn't need to redefine `A.render()` method at all (because the superclass's `render()` method does all the job for this already).